### PR TITLE
DataExporter: Enhance DataExporter to support other components #4414

### PR DESCRIPTION
**Pull request by ravihariharan**
DataExporter: Enhance DataExporter to support other components #4414 

Closed Old PR #4416

### CODE REVIEW

**Comment by Rapster** [comment]
Please don't change the format

**Comment by ravihariharan** [comment]
@Rapster  which format ? i cherry picked the commit and merge the code in a new branch. I also had to fix the License issue.

**Comment by Rapster** [comment]
See `Exporter#export()` parameters were previously well formatted

**Comment by Rapster** [comment]
I guess exporters should be renamed then, for instance `ExcelExporter` -> `ExcelDataTableExporter`?

**Comment by ravihariharan** [comment]
> I guess exporters should be renamed then, for instance `ExcelExporter` -> `ExcelDataTableExporter`?

The current changes are backward compatible. 

Since ExcelExporter and other were not final class i did not rename them.  
An example of breaking change would be extending ExcelExporter  and using a customExporter.
https://github.com/primefaces/primefaces/issues/4397

**Comment by Rapster** [comment]
Sorry I don't understand your point

**Comment by ravihariharan** [comment]
@Rapster  I made only the minimum changes that were required to support extending customExporter to be used with other components.  

We can rename all the class or 
We can move the DataTable specific classes to subpackage of .export.datatable
image

Lets see what @tandraschko / @melloware / @mertsincan   has to say about rename or moving these classes.

**Comment by Rapster** [comment]
+1 for creating the appropriate package, but I don't think all of them need to move in that new package. Few of these classes can be reused no matter what the component is

**Comment by ravihariharan** [comment]
> +1 for creating the appropriate package, but I don't think all of them need to move in that new package

All the class in the sub-package are specific to DataTable. (moved *Options class back)
image

ExporterFactory can be renamed to DataTableExporterFactory.

**Comment by Rapster** [comment]
Why callbacks and factory should be different from the rest? (This is a trick question btw :yum: )

**Comment by ravihariharan** [comment]
> Why callbacks and factory should be different from the rest?



image


We will have to remove <T extends UIComponent> from Exporter.
Will have to change the methods to have UIComponent  and cast when required.

Let me make few changes for factory  and in another branch and confirm.

**Comment by Rapster** [comment]
Well, nothing prevents you to create a `PDFExporter` interface, your "PDFDataTableExporter" implements it and that should solve your problem no?

**Comment by melloware** [comment]
I like all these changes.  And it will make future flexibility better.

**Comment by tandraschko** [comment]
+1 for the sub-package
and we should also name e.g. XMLExporter -> XMLDataTableExporter

**Comment by ravihariharan** [comment]
There are few discussion in https://github.com/primefaces/primefaces/issues/4414

@Rapster  approach was
Keep in mind, that my objective is to be able to reuse these exporters by doing so:
<p:dataExporter type="xls" target="tbl" fileName="cars" pageOnly="true"/>


I was going in the direction of using customExporter for other components. 

e.g. if we want to export a quickbooks format then we could do something like
<p:dataExporter type="quickbooks" customExporter="#{quickbooksExporter}" />
type="quickbooks" is only added for readability


i did not plan for a generic export for treetable , but allow developer to use the customExporter to implement the needed feature.

**Comment by ravihariharan** [comment]
image


CSVExporter ,  ExcelExporter , PDFExporter are now interface that are referred with corresponding VisitCallback

**Comment by ravihariharan** [comment]
@tandraschko / @melloware  / @Rapster i have made all the changes as discussed above. Please review and let me know if we can improve something.

1 task for future would be to create a Exporter factory.

**Comment by melloware** [comment]
This looks great to me.

**Comment by Rapster** [comment]
Since a package `datatable` has been created, I would have probably keep the original name. It's a lot of datatable-something in the datatable package if you know what I mean. Otherwise looks good

**Comment by ravihariharan** [comment]
This is what i would like to do @Rapster 
1. Move the datatable specific to org.primefaces.component.**datatable.export** package.
2. Rename the files back

Only reference in org.primefaces.component.export.DataExporter will be to the DataTableExporterFactory

image


i have the changes in another branch and will merge on confirmation :)

**Comment by Rapster** [comment]
I like your second suggestion in terms of packages. Indeed we can't have CSVExporter twice, and we can't use fully qualified class name... Let's keep `DataTable` as a prefix

File: src/main/java/org/primefaces/component/export/datatable/DataTableExporterBase.java [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+import javax.faces.component.html.HtmlGraphicImage;
+import javax.faces.context.FacesContext;
+import javax.faces.convert.Converter;
+
+import org.primefaces.component.celleditor.CellEditor;
+import org.primefaces.component.datatable.DataTable;
+import org.primefaces.component.overlaypanel.OverlayPanel;
+import org.primefaces.util.ComponentUtils;
+
+public abstract class DataTableExporterBase {
```
Comment (Rapster) [inline src/main/java/org/primefaces/component/export/datatable/DataTableExporterBase.java]: 1. Make this implements `Exporter<DataTable>` (and remove this from its children)
2. Rename into `DataTableExporter` (xxxBase is for definition of UIComponent)`

File: src/main/java/org/primefaces/component/export/datatable/DataTableCSVExporter.java [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+                    case FOOTER:
+                        textValue = col.getExportFooterValue() != null ? col.getExportFooterValue() : col.getFooterText();
+                        break;
+                    default:
+                        textValue = null;
+                        break;
+                }
+
+                if (textValue != null) {
+                    this.addColumnValue(builder, textValue);
```
Comment (Rapster) [inline src/main/java/org/primefaces/component/export/datatable/DataTableCSVExporter.java]: Remove unnecessary `this` references

File: src/main/java/org/primefaces/component/export/datatable/DataTableExporterBase.java (hunk 1 above)
Comment (ravihariharan) [reply src/main/java/org/primefaces/component/export/datatable/DataTableExporterBase.java]: Rename into DataTableExporter (xxxBase is for definition of UIComponent) - done

Comment (ravihariharan) [reply src/main/java/org/primefaces/component/export/datatable/DataTableExporterBase.java]: > Make this implements Exporter<DataTable> (and remove this from its children)

even if we do it , we will need the the children to implement specific interface.
e.g.

`public class DataTablePDFExporter extends DataTableExporter implements PDFExporter<DataTable>`

PDFExporter methods are reference from PDFExportVisitCallback

Comment (Rapster) [reply src/main/java/org/primefaces/component/export/datatable/DataTableExporterBase.java]: I know but still, DataTableExporter is an exporter (abstract but still an exporter :wink:)

**Comment by ravihariharan** [comment]
@Rapster / @tandraschko  PR ready for merge.

**Comment by Rapster** [comment]
I'm cool with it, do you think by any chance you could provide an example in the showcase?

**Comment by tandraschko** [comment]
@ravihariharan please also update the documentation (with samples how to use it)

**Comment by ravihariharan** [comment]
Migration Guide updated

**Comment by ravihariharan** [comment]
https://github.com/primefaces/showcase-facelift/pull/54

I reused a existing image to use for export for text.

image

**Pull request merged**